Compress CiliumEndpointSlices when moving from Identity to FCFS

dlapcevic opened the pull request:
CiliumEndpointSlice compression is required when changing the batching
mode from Identity to FCFS (First come, first served), to ensure the
expected performance of FCFS, which is achieved when CiliumEndpointSlices
are fully packed, instead of distributed by Identities.

_Example:_
Using many small K8s deployments might cause slow propagation of
CilumEndpointSlices while having Identity batching mode on. The reason
is that there are many CiliumEndpointSlices that need to be distributed
to all of the nodes, while there is rate limiting of CiliumEndpointSlice
updates. Switching to FCFS batching mode allows the same information
about CiliumEndpoints to be propagated through the cluster much faster, as
the number of slices will be reduced, since FCFS uses fully packed
slices. This change covers the compression of CiliumEndpointSlices in
case of changing the batching mode from Identity to FCFS.

Signed-off-by: Dorde Lapcevic [dordel@google.com](mailto:dordel@google.com)

alan-kut commented on operator/pkg/ciliumendpointslice/endpointslice.go:
```diff
@@ -169,6 +169,11 @@ func (c *CiliumEndpointSliceController) Run(ces cache.Indexer, stopCh chan struc
 		defer utilruntime.HandleCrash()
 	}()
 
+	if c.slicingMode != cesIdentityBasedSlicing {
```
Can it be generalized?

There are 4 cases:
Greedy -> Greedy
Greedy -> Identity
Identity -> Greedy
Identity -> Identity

This handles two cases for -> Greedy.

Don't we need anything like this for -> Identity?
If there is some other part of code that handles -> Identity can they be joined together?

dlapcevic replied:
The Identity batching will rearrange the CEPs in CESs accordingly on start up by design, while the Greedy (FCFS) won't. The two modes just have different requirements for the way they operate.

We can investigate more, but I've not seen a simple way to do it.

alan-kut commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -45,6 +47,7 @@ type operations interface {
 	// External APIs to Insert/Remove CEP in local dataStore
 	InsertCEPInCache(cep *cilium_v2.CoreCiliumEndpoint, ns string) string
 	RemoveCEPFromCache(cepName string, baseDelay time.Duration)
+	CompressCESsOnFCFSStartUp()
```
Why do we need to have it in 2 places?

dlapcevic replied:
This is just added to the operations interface definition, so that it can be called elsewhere.
The manager is exposed in the controller through operations interface.
https://github.com/cilium/cilium/blob/master/operator/pkg/ciliumendpointslice/endpointslice.go#L66

alan-kut commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -368,6 +371,110 @@ func (c *cesMgr) RemoveCEPFromCache(cepName string, baseDelay time.Duration) {
+
+	totalCEPs := 0
+	for _, ces := range cesList {
+		ces.backendMutex.RLock()
```
This code either doesn't need to be locked or you will have problems.

Why do you lock to get number of enpoints?

dlapcevic replied:
It's the same method used in the getTotalCEPCount() for the cached CESs.
https://github.com/cilium/cilium/blob/master/operator/pkg/ciliumendpointslice/manager.go#L372

I use a separate list here, but they still point to existing CESs (cesTrackers) that could be changing, and the lock is used to avoid incorrect read during the change.

alan-kut commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -368,6 +371,110 @@ func (c *cesMgr) RemoveCEPFromCache(cepName string, baseDelay time.Duration) {
+	}
+
+	// Calculate how many fully packed CES are needed.
+	requiredCESs := int(math.Ceil(float64(totalCEPs) / float64(c.maxCEPsInCES)))
```
Are changes to CESes possible when this code run?

dlapcevic replied:
Yes.
It will not affect the outcome, as when it tries to move CEPs from one CES, it uses the same method as if it was inserted, and it would just create a new CES to add it to.

alan-kut commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -368,6 +371,110 @@ func (c *cesMgr) RemoveCEPFromCache(cepName string, baseDelay time.Duration) {
+	// Calculate how many fully packed CES are needed.
+	requiredCESs := int(math.Ceil(float64(totalCEPs) / float64(c.maxCEPsInCES)))
+	// No compression is required when the number of CESs will remain the same.
+	if requiredCESs >= len(cesList) {
```
Can it be higher?

dlapcevic replied:
Although very unlikely, I think it's possible if there are many changes happening during this operation, since the CES controller's work queues are already processing events.

alan-kut commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -368,6 +371,110 @@ func (c *cesMgr) RemoveCEPFromCache(cepName string, baseDelay time.Duration) {
+	// Sort current list of CESs by size in decreasing order, to be able to take
+	// the largest ones and ensure the lowest number of CES updates.
+	sort.Slice(cesList, func(i, j int) bool {
+		return len(cesList[i].ces.Endpoints) > len(cesList[j].ces.Endpoints)
```
You don't lock here as you did above

dlapcevic replied:
I missed it.
Added.

dlapcevic committed:
Compress CiliumEndpointSlices when moving from Identity to FCFS

CiliumEndpointSlice compression is required when changing the batching
mode from Identity to FCFS (First come, first served), to ensure the
expected performance of FCFS, which is achieved when CiliumEndpointSlices
are fully packed, instead of distributed by Identities.

Example:
Using many small K8s deployments might cause slow propagation of
CilumEndpointSlices while having Identity batching mode on. The reason
is that there are many CiliumEndpointSlices that need to be distributed
to all of the nodes, while there is rate limiting of CiliumEndpointSlice
updates. Switching to FCFS batching mode allows the same information
about CiliumEndpoints to be propagated through the cluster much faster, as
the number of slices will be reduced, since FCFS uses fully packed
slices. This change covers the compression of CiliumEndpointSlices in
case of changing the batching mode from Identity to FCFS.

Signed-off-by: Dorde Lapcevic dordel@google.com
